### PR TITLE
[c++] NDArray shape

### PR DESCRIPTION
**Pull request by daniel-goldstein**
- Get shape for an `NDArrayExpression`. While this should (and in some cases does) **not** actually perform computations like element wise maps or matmul's, it doesn't in all cases because not everything is worked into being deforested. I'm going to take a closer look at deforesting/shape calculation and tie up loose ends, but there are some subtleties and I think this is a good chunk.
- Changed IR nodes that are concerned with shape (`MakeNDArray`, `NDArrayReshape`) take shapes as hail tuples (instead of Scala sequences of IR, or IR with another parameter for length). The only thing Scala sequences were really useful for were statically knowing the number of dimensions which we can get with tuples.

### CODE REVIEW

File: hail/src/main/scala/is/hail/cxx/Emit.scala [hunk 1]
```diff
@@ -876,60 +880,78 @@ class Emitter(fb: FunctionBuilder, nSpecialArgs: Int, ctx: SparkFunctionContext)
              | ${ rowMajort.setup }
              | ${ shapet.setup }
              | ${ datat.setup }
-             | if (${ rowMajort.m } || ${ shapet.m } || ${ datat.m }) {
+             | ${ shapeTup.define }
+             | if (${ datat.m } || ${ rowMajort.m } || ${ shapeMissing.foldRight("false")((b, m) => s"$b || $m") }) {
```
Comment (patrick-schultz) [inline hail/src/main/scala/is/hail/cxx/Emit.scala]: I think this still needs to check `shapet.m` before looking at `shapeMissing`, no?

File: hail/src/main/scala/is/hail/cxx/Emit.scala [hunk 2]
```diff
@@ -876,60 +880,78 @@ class Emitter(fb: FunctionBuilder, nSpecialArgs: Int, ctx: SparkFunctionContext)
...
         val shape = fb.variable("shape", "std::vector<long>")
-        val buildShape = shapet.zip(shapeVars).map { case (lent, lenVar) =>
+        val buildShape = Seq.tabulate(shapePType.size) { dim =>
           s"""
-             | if (${ lent.m }) {
+             | if (${ shapePType.cxxIsFieldMissing(shapeTup.toString, dim) }) {
              |  ${ fb.nativeError("Cannot reshape with missing dimension length") }
              | }
-             | $shape.push_back($lenVar);
+             | $shape.push_back(${ shapePType.cxxLoadField(shapeTup.toString, dim) });
```
Comment (patrick-schultz) [inline hail/src/main/scala/is/hail/cxx/Emit.scala]: Same comment about building the shape vector.

File: hail/src/test/scala/is/hail/expr/ir/IRSuite.scala [hunk 3]
```diff
@@ -986,18 +991,17 @@ class IRSuite extends SparkSuite {
     val shape = MakeArray(FastSeq(2L, 5L).map(I64), TArray(TInt64()))
```
Comment (patrick-schultz) [inline hail/src/test/scala/is/hail/expr/ir/IRSuite.scala]: Doesn't this need to be a tuple?

File: hail/python/test/hail/test_ir.py [hunk 4]
```diff
@@ -53,8 +53,7 @@ def value_irs(self):
             ir.ToSet(a),
             ir.ToDict(da),
             ir.ToArray(a),
-            ir.MakeNDArray(2,
-                           ir.MakeArray([ir.F64(-1.0), ir.F64(1.0)], hl.tarray(hl.tfloat64)),
+            ir.MakeNDArray(ir.MakeArray([ir.F64(-1.0), ir.F64(1.0)], hl.tarray(hl.tfloat64)),
                            ir.MakeArray([ir.I64(1), ir.I64(2)], hl.tarray(hl.tint64)),
```
[the remark this answers is not in the document]
Comment (daniel-goldstein) [reply hail/python/test/hail/test_ir.py]: yup! Missed that, I wonder how these are passing...

File: hail/src/main/scala/is/hail/cxx/Emit.scala [hunk 5]
```diff
@@ -855,17 +855,21 @@ class Emitter(fb: FunctionBuilder, nSpecialArgs: Int, ctx: SparkFunctionContext)
...
         val rowMajort = emit(rowMajorIR)
 
-        val elemSize = elemPType.byteSize
-        val shape = fb.variable("shape", "std::vector<long>",
-          s"load_non_missing_vector<${shapeContainer.cxxImpl}>(${shapet.v})")
+        val shape = fb.variable("shape", "std::vector<long>")
+        val shapeTup = fb.variable("shape_tuple", "const char *", shapet.v)
+        val shapeMissing = Seq.tabulate(shapePType.size) { shapePType.cxxIsFieldMissing(shapeTup.toString, _) }
+        val buildShape = Seq.tabulate(shapePType.size) { dim =>
+          s"$shape.push_back(${ shapePType.cxxLoadField(shapeTup.toString, dim) });"
```
[the remark this answers is not in the document]
Comment (daniel-goldstein) [reply hail/src/main/scala/is/hail/cxx/Emit.scala]: Good point

File: hail/src/main/scala/is/hail/cxx/Emit.scala (hunk 1 above)
Comment (daniel-goldstein) [reply hail/src/main/scala/is/hail/cxx/Emit.scala]: Ah right the whole thing could be missing

File: hail/python/test/hail/test_ir.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (patrick-schultz) [reply hail/python/test/hail/test_ir.py]: Should probably try to make sure some test fails before you fix it.

**Pull request merged**